src/lib/provable/test/merkle-list.test.ts

```diff
@@ -0,0 +1,99 @@
+import { Field, MerkleList, Poseidon } from 'o1js';
+
+describe('Merkle List and Iterator', () => {
+  describe('primitive data structures', () => {
+    const emptyHash = Poseidon.hash([Field(0)]);
+    const nextHash = (hash: Field, value: Field) =>
+      Poseidon.hash([hash, value]);
+    class List extends MerkleList.create(Field, nextHash, emptyHash) {}
+
+    describe('should initialize same hash', () => {
+      it('empty hash', () => {
+        let l1 = List.empty();
+        l1.hash.assertEquals(emptyHash);
+
+        let l2 = List.from([]);
+        l2.hash.assertEquals(emptyHash);
+      });
+    });
+
+    describe('should hash correctly', () => {
+      it('empty hash', () => {
+        let l1 = List.from([Field(1), Field(2)]);
+        let l1Hash = l1.hash;
+
+        let l2 = List.empty();
+        l2.push(Field(1));
+        l2.push(Field(2));
+        let l2Hash = l2.hash;
+
+        l1Hash.assertEquals(l2Hash);
+      });
+    });
+
+    describe('pop and previous should behave the same', () => {
+      it('empty hash', () => {
+        let l1 = List.from([Field(1), Field(2)]);
+        let l2 = l1.clone();
```

Comment:
is this cloning necessary 🤔

Reply:
yea, otherwise `pop` would literally steal the elements form the iterator :D I could have also just initiated a second array, but I thought I would clone it to make sure it actually does clone